src/interfaces/messages/handlers/messages-get.ts

```diff
@@ -0,0 +1,89 @@
+import type { DataStore } from '../../../store/data-store.js';
+import type { DidResolver } from '../../../did/did-resolver.js';
+import type { MessageStore } from '../../../store/message-store.js';
+import type { MethodHandler } from '../../types.js';
+import type { MessagesGetMessage, MessagesGetReply, MessagesGetReplyEntry } from '../types.js';
+
+import { DataStream } from '../../../utils/data-stream.js';
+import { DwnConstant } from '../../../core/dwn-constant.js';
+import { Encoder } from '../../../utils/encoder.js';
+import { MessageReply } from '../../../core/message-reply.js';
+import { MessagesGet } from '../messages/messages-get.js';
+import { authenticate, authorize } from '../../../core/auth.js';
+import { DwnInterfaceName, DwnMethodName, Message } from '../../../core/message.js';
+
+type HandleArgs = { tenant: string, message: MessagesGetMessage };
+
+export class MessagesGetHandler implements MethodHandler {
+  constructor(private didResolver: DidResolver, private messageStore: MessageStore, private dataStore: DataStore) {}
+
+  public async handle({ tenant, message }: HandleArgs): Promise<MessagesGetReply> {
+    let messagesGet: MessagesGet;
+
+    try {
+      messagesGet = await MessagesGet.parse(message);
+    } catch (e) {
+      return MessageReply.fromError(e, 400);
+    }
+
+    try {
+      await authenticate(message.authorization, this.didResolver);
+      await authorize(tenant, messagesGet);
+    } catch (e) {
+      return MessageReply.fromError(e, 401);
+    }
+
+    const promises: Promise<MessagesGetReplyEntry>[] = [];
+
+    for (const messageCid of message.descriptor.messageCids) {
+      const promise = this.messageStore.get(tenant, messageCid)
+        .then(message => {
+          return { messageCid, message };
+        })
+        .catch(_ => {
+          // FIXIT: log error? throw error?
+          return { messageCid, message: undefined };
+        });
+
+      promises.push(promise);
+    }
+
+    const messages = await Promise.all(promises);
```

Comment:
NIT: instead of creating a `promises` array you could do this with `Array.prototype.map`

and also at the same time you could merge the `encodedData` part so that we dont iterate over all the `messages` twice

```
const messages = await Promise.all(message.descriptor.messageCids.map(async messageCid => {
  const message = await this.messageStore.get(tenant, messageCid);
  if (!message) {
    return { messageCid, error }; // FIXME
  }

  if (message.descriptor.interface === DwnInterfaceName.Records && message.descriptor.method === DwnMethodName.Write) {
    const dataCid = message.descriptor.dataCid;
    const dataSize = message.descriptor.dataSize;
    if (dataCid !== undefined && dataSize! <= DwnConstant.maxDataSizeAllowedToBeEncoded) {
      const messageCid = await Message.getCid(message);
      const { dataStream } = await dataStore.get(tenant, messageCid, dataCid) ?? { };
      if (dataStream) {
        const dataBytes = await DataStream.toBytes(dataStream);
        message.encodedData = Encoder.bytesToBase64Url(dataBytes);
      }
    }
  }

  return { messageCid, message };
});
```